File: 030_substring_with_concatenation_of_all_words/map/Solution.java
import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;

class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> results = new ArrayList<>();
        int strLen = s.length();
        int arrLen = words.length;
        if (strLen == 0 || arrLen == 0) return results;
        int wordLen = words[0].length();

        HashMap<String, Integer> wordMap = new HashMap<>();
        for (String word : words) {
            int value = wordMap.getOrDefault(word, 0);
            wordMap.put(word, value + 1);
        }

        HashMap<String, Integer> curMap = new HashMap<>();
        int count = 0;
        for (int i = 0; i < strLen - arrLen * wordLen + 1; i++) {
            while (count < arrLen) {
                String curWord = s.substring(i + count * wordLen, i + (count + 1) * wordLen);
                if (wordMap.containsKey(curWord)) {
                    int value = curMap.getOrDefault(curWord, 0);
                    if (value + 1 > wordMap.get(curWord)) break;
                    curMap.put(curWord, value + 1);
                } else {
                    break;
                }
                count++;
            }
            if (count == arrLen) results.add(i);
            curMap.clear();
            count = 0;
        }

        return results;
    }

    public static void main(String[] args) {
        String s = "barfoothefoobarman";
        String[] words = new String[]{"foo", "bar"};
        Solution solution = new Solution();
        List<Integer> results = solution.findSubstring(s, words);
        for (int result : results) {
            System.out.print(result + " ");
        }
    }
}

File: 030_substring_with_concatenation_of_all_words/map/Solution.py
class Solution:
    def findSubstring(self, s: str, words: list) -> list:
        results = []
        s_len = len(s)
        arr_len = len(words)
        if s_len == 0 or arr_len == 0: return results
        word_len = len(words[0])

        word_map = {}
        for word in words:
            value = word_map.get(word)
            word_map[word] = value + 1 if value else 1

        for i in range(s_len - word_len * arr_len + 1):
            cur_map = {}
            count = 0
            while count < arr_len:
                cur_word = s[i + count * word_len: i + (count + 1) * word_len]
                if cur_word in word_map.keys():
                    value = cur_map.get(cur_word)
                    if value and value + 1 > word_map.get(cur_word): break
                    cur_map[cur_word] = value + 1 if value else 1
                else: break
                count += 1
            if count == arr_len: results.append(i)

        return results


if __name__ == '__main__':
    s = 'barfoothefoobarman'
    words = ['foo', 'bar']
    solution = Solution()
    print(solution.findSubstring(s, words))

